### PR TITLE
Background image: ensure consistency with defaults and fix reset/remove functionality

### PR DESCRIPTION
**Pull request by ramonjd**
Follow up to:

- https://github.com/WordPress/gutenberg/pull/64192

## What and how?

Block background styles will be part of WordPress 6.7.

This PR fixes miscellaneous bugs in the background image default values and controls. It:

- ensures that inherited values are shown in the controls
- ensures that any default values for **uploaded images** are represented in the controls
- ~~blocks should honour any inherited styles. That means, defaults should not overwrite styles defined in theme.json or global styles~~
- changes the default values of "center" to "50% 50%" so that it is represented in the focus picker control values. 
- ensures "resetting" an block's styles resets any current styles (at the moment it does not)
- ensures "removing" an image resets any current styles and remove access to the background controls

## Why?

To address a lot of inconsistencies in the way defaults are applied to block background styles.

## Testing Instructions

This might take some time, sorry!

Using a theme with background images set on blocks (see example theme.json), add some blocks to the editor (see example blocks).

Ensure that you can see inherited styles in the controls and that no background default values are applied.

Now upload/add your own images to these blocks.

1. The background size value should be set to "cover"
2. When toggling between background sizes, the defaults should be visible, E.g., Switch to "contain" and the image should be centered ("50% 50%"), and the value reflected in the controls.
3. Save and reset the image. Make sure that all values on the image have been reset.
4. Change the background size or other value for the inherited image. 
5. Now, "remove" the image so that the background is none. All properties, expect "background-image:none" should be gone.

Ensure that the frontend matches what you're seeing in the editor.

Site-wide background images should operate as before and be unaffected by any defaults aside from the existing position of `50% 0` position when switching to tiled.

<details>

<summary>Example block HTML</summary>
```html
<!-- wp:quote {"style":{"background":{"backgroundImage":"none"}}} -->
<blockquote class="wp-block-quote"><!-- wp:paragraph {"style":{"color":{"text":"#f8f8f8"},"elements":{"link":{"color":{"text":"#f8f8f8"}}},"typography":{"fontSize":"51px"}}} -->
<p class="has-text-color has-link-color" style="color:#f8f8f8;font-size:51px">Quote</p>
<!-- /wp:paragraph --></blockquote>
<!-- /wp:quote -->

<!-- wp:verse {"style":{"elements":{"link":{"color":{"text":"var:preset|color|white"}}},"typography":{"fontSize":"61px"},"background":{"backgroundImage":{"url":"http://localhost:8888/wp-content/uploads/2024/08/rampant-onion.jpg","id":24,"source":"file","title":"rampant-onion"},"backgroundSize":"cover"}},"textColor":"white"} -->
<pre class="wp-block-verse has-white-color has-text-color has-link-color" style="font-size:61px">Verse</pre>
<!-- /wp:verse -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"text":"#fefefe"},"elements":{"link":{"color":{"text":"#fefefe"}}},"typography":{"fontSize":"59px"}}} -->
<p class="has-text-color has-link-color" style="color:#fefefe;font-size:59px">Group</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```
</details>

<details>

<summary>example theme.json</summary>

```json

{
	"$schema": "../../schemas/json/theme.json",
	"version": 3,
	"settings": {
		"appearanceTools": true,
		"layout": {
			"contentSize": "840px",
			"wideSize": "1100px"
		}
	},
	"styles": {

		"background": {
			"backgroundImage": {
				"url": "https://images.pexels.com/photos/105819/pexels-photo-105819.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr="
			}
		},
		"blocks": {
			"core/verse": {
				"background": {
					"backgroundImage": {
						"url": "https://images.pexels.com/photos/27332202/pexels-photo-27332202/free-photo-of-surfer.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
					},
					"backgroundSize": "contain"
				},
				"dimensions": {
					"minHeight": "100px"
				}
			},
			"core/quote": {
				"background": {
					"backgroundImage": {
						"url": "https://images.pexels.com/photos/22484288/pexels-photo-22484288/free-photo-of-the-circular-stone-terraces-of-the-inca-ruins.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
					}
				},
				"dimensions": {
					"minHeight": "100px"
				}
			},
			"core/group": {
				"background": {
					"backgroundImage": {
						"url": "https://images.pexels.com/photos/27101553/pexels-photo-27101553/free-photo-of-a-cocktail-with-mint-and-lime-in-it.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
					}
				},
				"dimensions": {
					"minHeight": "111px"
				}
			}
		}
	}
}

```

</details>

## Screenshots or screencast 

https://github.com/user-attachments/assets/8e04f760-9811-42e7-8a2a-dca698aa73d9

## TODO

- [ ] The UI is getting complicated. We probably need E2E tests in a follow up.

### CODE REVIEW

File: lib/block-supports/background.php [hunk 1]
```diff
@@ -58,11 +58,18 @@ function gutenberg_render_background_support( $block_content, $block ) {
...
 	$background_styles['backgroundAttachment'] = $block_attributes['style']['background']['backgroundAttachment'] ?? null;
-	if ( ! empty( $background_styles['backgroundImage'] ) ) {
-		$background_styles['backgroundSize'] = $background_styles['backgroundSize'] ?? 'cover';
+	if ( isset( $background_styles['backgroundImage']['id'] ) ) {
+		$inherited_styles                  = gutenberg_get_global_styles();
+		$inherited_block_background_styles = $inherited_styles['blocks'][ $block['blockName'] ]['background'] ?? null;
+		$inherited_background_size         = $inherited_block_background_styles['backgroundSize'] ?? null;
+		if ( ! isset( $background_styles['backgroundSize'] ) && ! $inherited_background_size ) {
+			$background_styles['backgroundSize'] = 'cover';
+		}
```
Comment (andrewserong) [inline lib/block-supports/background.php]: Apologies if I'm missing something, but just thinking about this one a little more. Why do we need to check the inherited styles in the server-rendered output of the block instance level styles?

In this case, the image will always be one that the user set (or that's bundled in a pattern), so in that case, won't we want to go with the defaults rather than styles inherited from global styles? I.e. if a user sets an image at the block instance level, we assume that they're setting things in the UI to their liking rather than inheriting from global styles at that point 🤔

Mostly I think my question boils down to: do we need to complexity of this look up? In the UI at least we kind of bundle setting an image with all the other properties now, so I wasn't sure how necessary it is to allow this inheritance.

Edit: on the other hand I _do_ see the value in attempting to preserve what's set in global styles so that it overwrites our defaults logic! I suppose I'm torn on this one 😄

Comment (ramonjd) [reply lib/block-supports/background.php]: > Why do we need to check the inherited styles in the server-rendered output of the block instance level styles?

Good question. It's all very complex, you're right.

My assumption on this one is that, where the theme or global styles provide values for a block, the block should use them if no local styles exist to overwrite them.

So if a theme defines "contain" for a block but nothing else, and a user uploads a new image to that block, whether in global styles or at the block level, the default will be the same as if they chose "contain" themselves.

The other, and probably easier option, would be to just set "cover" every time a user adds an image and forget about inherited values?

File: lib/class-wp-theme-json-gutenberg.php [hunk 2]
```diff
@@ -2391,7 +2391,7 @@ protected static function compute_style_properties( $styles, $settings = array()
 					$styles['background']['backgroundSize'] = $styles['background']['backgroundSize'] ?? 'cover';
 					// If the background size is set to `contain` and no position is set, set the position to `center`.
 					if ( 'contain' === $styles['background']['backgroundSize'] && empty( $styles['background']['backgroundPosition'] ) ) {
-						$styles['background']['backgroundPosition'] = 'center';
+						$styles['background']['backgroundPosition'] = '50% 50%';
```
Comment (ramonjd) [inline lib/class-wp-theme-json-gutenberg.php:2394]: Changing from "center" to "50% 50%" because it's the same value, but we can display this in the controls.

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 3]
```diff
@@ -562,24 +594,27 @@ function BackgroundSizeControls( {
...
 
+	// Set a default background position for non-site-wide, uploaded images with a size of 'contain'.
+	const backgroundPositionValue =
+		! positionValue && isUploadedImage && 'contain' === sizeValue
+			? defaultValues?.backgroundPosition
+			: positionValue;
+
 	return (
-		<VStack spacing={ 4 } className="single-column">
+		<VStack spacing={ 3 } className="single-column">
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:586]: These changes are about reducing space, can be in a separate PR

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 4]
```diff
@@ -562,24 +594,27 @@ function BackgroundSizeControls( {
...
-				value={ backgroundPositionToCoords( positionValue ) }
+				value={ backgroundPositionToCoords( backgroundPositionValue ) }
 				onChange={ updateBackgroundPosition }
 			/>
 			<ToggleControl
 				__nextHasNoMarginBottom
 				label={ __( 'Fixed background' ) }
 				checked={ attachmentValue === 'fixed' }
 				onChange={ toggleScrollWithPage }
-				help={ __(
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:580]: Another change about reducing space, can be in a separate PR

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 5]
```diff
@@ -700,8 +735,13 @@ export default function BackgroundPanel( {
 		hasBackgroundImageValue( value ) ||
 		hasBackgroundImageValue( inheritedValue );
 
+	const imageValue =
+		value?.background?.backgroundImage ||
+		inheritedValue?.background?.backgroundImage;
+
 	const shouldShowBackgroundImageControls =
 		hasImageValue &&
+		'none' !== imageValue &&
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:726]: Ensure the panel closes where an image is removed

File: lib/block-supports/background.php (hunk 1 above)
Comment (andrewserong) [reply lib/block-supports/background.php]: I very much appreciate the attention to detail here! I think the way you've done is the _right_ way to go about it while handling a complex situation. But my gut feeling is that it's probably more complex than we need it to be now that we have all the controls bundled together in the popover where folks can make adjustments easily.

> The other, and probably easier option, would be to just set "cover" every time a user adds an image and forget about inherited values?

That sounds like a good option to try to me!

Comment (ramonjd) [reply lib/block-supports/background.php]: @andrewserong I've simplified this PR and focussed on bug/inconsistency fixing. 

Thank you for helping me move it forward! 🙇🏻

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 6]
```diff
@@ -372,7 +376,9 @@ function BackgroundImageControls( {
 
 	const onRemove = () =>
 		onChange(
-			setImmutably( style, [ 'background', 'backgroundImage' ], 'none' )
+			setImmutably( style, [ 'background' ], {
+				backgroundImage: 'none',
+			} )
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:381]: Here all background styles are wiped except "none".

The reason is that it should behave as though there is no image, and replacing the image shouldn't inherit any previously-set styles.

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 7]
```diff
@@ -469,11 +474,19 @@ function BackgroundSizeControls( {
...
+	let currentValueForToggle =
 		! sizeValue && isUploadedImage
 			? defaultValues?.backgroundSize
 			: sizeValue || 'auto';
-
+	/*
+	 * The incoming value could be a value + unit, e.g. '20px'.
+	 * In this case set the value to 'tile'.
+	 */
+	currentValueForToggle = ! [ 'cover', 'contain', 'auto' ].includes(
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:485]: This is a new check to activate the "Tile" toggle for non-word values, e.g., value + unit. 

This ensure consistency with toggling sizes, where clicking on tile and updating the width will not deactivate tile - the "tile" effect is still there as far as the user is concerned.

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 8]
```diff
@@ -510,6 +523,7 @@ function BackgroundSizeControls( {
 			 * receives a default background position of '50% 0',
 			 * when the toggle switches to "Tile". This is to increase the chance that
 			 * the image's focus point is visible.
+			 * This is in-editor only to assist with the user experience.
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:526]: More comments just to explain all this madness.

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 9]
```diff
@@ -562,24 +576,27 @@ function BackgroundSizeControls( {
 			)
 		);
 
+	// Set a default background position for non-site-wide, uploaded images with a size of 'contain'.
+	const backgroundPositionValue =
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:580]: Display the defaults if any are set.

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 10]
```diff
@@ -725,7 +747,7 @@ export default function BackgroundPanel( {
 				) }
 			>
 				<ToolsPanelItem
-					hasValue={ () => hasImageValue }
+					hasValue={ () => !! value?.background }
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:750]: Because we can change size, repeat etc values of inherited background images, we should check for all properties, not just the image to indicate whether the controls can be reset.

File: packages/block-editor/src/hooks/background.js [hunk 11]
```diff
@@ -55,31 +56,28 @@ export function hasBackgroundSupport( blockName, feature = 'any' ) {
 }
 
 export function setBackgroundStyleDefaults( backgroundStyle ) {
-	if ( ! backgroundStyle ) {
+	if ( ! backgroundStyle || ! backgroundStyle?.backgroundImage?.url ) {
```
Comment (ramonjd) [inline packages/block-editor/src/hooks/background.js:59]: Just tightening things up here.

**Review by andrewserong — APPROVED** [review]
Just walked through this code change and demo with @ramonjd on a call, and this is looking terrific to me!

✅ Works nicely at the block level in post content
✅ Block level in global styles is working correctly
✅ As-is site wide background images

LGTM! 🚀

**Comment by jorgefilipecosta** [comment]
I just cherry-picked this PR to the release/19.0 branch to get it included in the next release: 152194662e7
